src/test/java/org/opentripplanner/calendar/impl/CalendarServiceDataFactoryImplTest.java

```diff
@@ -0,0 +1,214 @@
+        LocalizedServiceId locServiceId = getWeekdaysLocalizedServiceId();
+
+        assertEquals("ServiceId(id=F_weekdays timeZone=America/New_York)", locServiceId.toString());
+        assertEquals(2, data.getLocalizedServiceIds().size());
+    }
+
+    @Test
+    public void testDataGetDatesForLocalizedServiceId() throws IOException {
+        LocalizedServiceId locServiceId = getWeekdaysLocalizedServiceId();
+
+        SimpleDateFormat formatDate = new SimpleDateFormat("yyyy-MM-dd", Locale.FRENCH);
+
+        assertEquals("[2009-01-01, 2009-01-02, 2009-01-05, 2009-01-06, 2009-01-07]",
+                data.getDatesForLocalizedServiceId(locServiceId).stream().limit(5)
+                        .map(formatDate::format).collect(toList()).toString());
+    }
+
+    @Test
+    public void testServiceGetServiceIdsOnDate() throws IOException {
+        Set<FeedScopedId> servicesOnFriday = calendarService.getServiceIdsOnDate(A_FRIDAY);
+        assertEquals("[F_alldays, F_weekdays]", sort(servicesOnFriday).toString());
+
+        Set<FeedScopedId> servicesOnSunday = calendarService.getServiceIdsOnDate(A_SUNDAY);
+        assertEquals("[F_alldays]", servicesOnSunday.toString());
+
+        // Test exclusion of serviceCalendarDate
+        Set<FeedScopedId> servicesOnMonday = calendarService.getServiceIdsOnDate(A_MONDAY);
+        assertEquals("[F_weekdays]", servicesOnMonday.toString());
+    }
+
+    @Test
+    public void testServiceGetServiceIds() throws IOException {
+        Set<FeedScopedId> serviceIds = calendarService.getServiceIds();
+        assertEquals("[F_alldays, F_weekdays]", sort(serviceIds).toString());
+    }
+
+    @Test
+    public void testServiceGetTimeZoneForAgencyId() throws IOException {
+        TimeZone result = calendarService.getTimeZoneForAgencyId(AGENCY);
+        assertEquals("America/New_York", result.getID());
+    }
+
+    @Test
+    public void testServiceGetServiceDatesForServiceId() throws IOException {
+        Set<ServiceDate> alldays = calendarService.getServiceDatesForServiceId(SERVICE_ALLDAYS_ID);
+
+        assertTrue(alldays.contains(A_FRIDAY));
+        assertTrue(alldays.contains(A_SUNDAY));
+        assertEquals(14975, alldays.size());
+
+        Set<ServiceDate> weekdays = calendarService
+                .getServiceDatesForServiceId(SERVICE_WEEKDAYS_ID);
+
+        assertTrue(weekdays.contains(A_FRIDAY));
+        assertFalse(weekdays.contains(A_SUNDAY));
+        assertEquals(10697, weekdays.size());
+    }
+
+    private LocalizedServiceId getWeekdaysLocalizedServiceId() {
+        Set<LocalizedServiceId> localizedServiceIds = data.getLocalizedServiceIds();
+        return localizedServiceIds.stream().filter(it -> it.getId().equals(SERVICE_WEEKDAYS_ID))
+                .findFirst().orElseThrow(IllegalStateException::new);
+    }
+
+    private static GtfsContext createCtxBuilder() throws IOException {
+        GtfsContextBuilder contextBuilder = contextBuilder(FEED_ID, ConstantsForTests.FAKE_GTFS);
```

Comment:
This is extracting a nested builder out of another partially initialized builder, then acting upon the nested builder before calling build() on the outer builder. It seems to me that this should be refactored if possible such that we do not "reach inside" builder elements that should be encapsulated and manipulate them. Perhaps we can first completely build the OtpTransitService, then supply that to the GtfsContextBuilder. But then I see that the GtfsContext is actually returned containing an unfinished OtpTransitServiceBuilder. It all feels quite ad hoc. Again I see how this is derived from the existing code, which was not well organized to begin with, but I hope we can eventually refactor it to be cleaner.

Reply:
Agree - I think the refactoring make problem visible - hence we now have the tool to do something with it.